sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1817,26 +1817,32 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
 
     case filter @ Filter(condition, union: Union) =>
       // Union could change the rows, so non-deterministic predicate can't be pushed down
+      val output = union.output
       val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition(_.deterministic)
 
       if (pushDown.nonEmpty) {
         val pushDownCond = pushDown.reduceLeft(And)
-        val output = union.output
```

Comment:
super nit: seems better to keep this local variable here instead of moving it up.

Reply:
Maybe, I was looking at union, output is a slightly non-trivial computation and we look into it many times?

Reply:
Then we can save one computation by putting the local variable inside the `if` branch as previous.

Reply:
Sounds good :)